build.gradle

```diff
@@ -190,27 +190,60 @@ task cmakeBuild(type: Exec) {
     ]
 }
 
-// run jtreg tests. Note: needs jtreg_home variable set to point to the jtreg
-task jtreg(type: JavaExec) {
-    dependsOn createRuntimeImageForTest,cmakeBuild
+void createJtregTask(String name, boolean coverage, String os_lib_dir) {
+    tasks.register(name, JavaExec) {
+        dependsOn createRuntimeImageForTest,cmakeBuild
 
-    doFirst {
         if (findProperty("jtreg_home") == null) {
             throw new GradleException("jtreg_home is not defined")
         }
+        // e.g.: <jacoco repo>/org.jacoco.agent/target/classes/jacocoagent.jar
+        if (coverage && findProperty("jacoco_agent") == null) {
+            throw new GradleException("jacoco_agent is not defined")
+        }
+
+        workingDir = "$buildDir"
+
+        classpath = files(findProperty("jtreg_home") + "/lib/jtreg.jar")
+
+        args = [
+                "-jdk", "$buildDir/jextract-jdk-test-image",
+                "-nativepath:$buildDir/testlib-install/${os_lib_dir}",
+                "-javaoption:--enable-preview",
+                "-javaoption:--enable-native-access=org.openjdk.jextract,ALL-UNNAMED",
+                "-avm", "-conc:auto", "-verbose:summary",
+                "-retain:fail,error",
+        ]
+
+        if (coverage) {
+            String jacocoAgent = findProperty("jacoco_agent")
+            String coverageFile = "$buildDir/jacoco-run/jextract.exec"
+            String includes = "org.openjdk.jextract.*"
+            args += "-javaoption:-javaagent:$jacocoAgent=destfile=$coverageFile,includes=$includes"
+        }
+
+        args += "../test"
     }
+}
 
-    workingDir = "$buildDir"
+createJtregTask("jtreg", false, os_lib_dir)
+createJtregTask("jtregWithCoverage", true, os_lib_dir)
+
+tasks.register("coverage", JavaExec) {
```

Comment:
Note that I'm using `tasks.register` here to make the configuration of this task lazy. That allows us to avoid an error when the `jacoco_cli` is not set, but the task isn't invoked.

The old jtreg task did this in a `doFirst` block, but that code actually doesn't run until the task itself is invoked. So, if there is a missing property, and the configuration references that property, we'll get an NPE instead of the intended exception.

Note that I've moved the property null check out of the `doFirst` block of the jtreg task as well, and made the configuration lazy there too. If I do that, I see an exception as expected when the task is configured, but not when the task doesn't run.